lib/promise_utils.js

```diff
@@ -36,9 +43,68 @@ function createDoneTask () {
   return task
 }
 
+/**
+ * Similar to the 'once' function from the 'events' module, but allows you to add a condition for when you want to
+ * actually handle the event, as well as a timeout. The listener is additionally removed if a timeout occurs, instead
+ * of with 'once' where a listener might stay forever if it never triggers.
+ * Note that timeout and checkCondition, both optional, are in the third parameter as an object.
+ * @param emitter - The event emitter to listen to
+ * @param event - The name of the event you want to listen for
+ * @param [timeout=0] - An amount, in milliseconds, for which to wait before considering the promise failed. <0 = none.
+ * @param [checkCondition] - A function which matches the same signature of an event emitter handler, and should return something truthy if you want the event to be handled. If this is not provided, all events are handled.
+ * @returns {Promise} A promise which will either resolve to an *array* of values in the handled event, or will reject on timeout if applicable. This may never resolve if no timeout is set and the event does not fire.
+ */
+function onceWithCleanup (emitter, event, { timeout = 0, checkCondition = undefined } = {}) {
+  const task = createTask()
+
+  const onEvent = (...data) => {
+    if (checkCondition && !checkCondition(...data)) {
+      return
+    }
+
+    task.finish(data)
+  }
+
+  emitter.addListener(event, onEvent)
+
+  if (timeout > 0) {
+    // For some reason, the call stack gets lost if we don't create the error outside of the .then call
+    const timeoutError = new Error(`Event ${event} did not fire within timeout of ${timeout}ms ${checkCondition && checkCondition.name}`)
+    sleep(timeout).then(() => {
+      if (!task.done) {
+        task.cancel(timeoutError)
+      }
+    })
+  }
+
+  task.promise.finally(() => emitter.removeListener(event, onEvent))
+
+  return task.promise
+}
+
+function withTimeout (promise, timeout) {
+  return new Promise((resolve, reject) => {
+    const timeoutHandle = setTimeout(() => reject(new Error('Promise timed out.')), timeout)
+    promise
+      .finally(() => {
+        clearTimeout(timeoutHandle)
+      })
+      .then((result) => {
+        resolve(result)
+        return result
```

Comment:
Why are you returning result or error there?

Reply:
In case anything else attaches to this promise, result/error need to be returned so that they can still use the result. Though there might not be any use cases for it in this PR, it is possible that someone may want to timeout on a promise in one place but handle it in parallel in another function.